lib/div-alg.gi

```diff
@@ -1706,25 +2209,599 @@ local W,l,W1,A,i;
+
+
+for x in [1..k] do
+  F2:=NF(con,[A[4][x][2]]);
+  if ForAll([1..k],j->e[x][j]=0) and IsCyclotomicExtension(F2,F) then
+    y := Difference([1..k],[x]);
+    F1:=NF(con,List(y,j->A[4][j][2]));
+    c := Lcm(2,Conductor(F1));
+    g := E(c);
+    while not g in F1 do
+      g:=g*E(c);
+    od;
+    h := E(con)^A[4][x][3];
+    n := Order(g);
+    split := false;
+    a := 1;
+    for i in [0..n-1] do
+      split := h=Norm(F1,F,a);
+      if split then
+        break;
+      fi;
+      a:=a*g;
+    od;
+## If not we check whether it is a cyclotomic and compute the Schur index to check whether it is 1
+    if not split and IsCyclotomicExtension(F1,F) then
+      a := 1;
+      ex := 0;
+      for i in [0..n-1] do
+        if a = h then
+          break;
+        else
+          ex:=ex+1;
+          a:=a*g;
+        fi;
+      od;
+      A1 := [1,F,c,[A[4][x][1],A[4][x][2] mod c,ex]];
+      split := SchurIndex(A1)=1;  
+    fi;
+    if split then
+      c := Lcm(2,Conductor(F2));
+      g:=E(c);
+      act := [];
+      l := Length(y);
+      pos := 1;
+      if l>1 then 
+        coc1 := [];    
+      fi;
+      for i in y do
+        a:=1;
+        h:=E(con)^A[4][i][3];
+        for ex in [0..c] do
+          if a = h then
+            break;
+          fi;
+          a:=a*g;
+        od;
+        if ex=c then 
+          Print("\n The algebra is not a genuine cyclotomic algebra \n");
+          return fail;
+        fi;          
+        Add(act,[A[4][i][1],A[4][i][2] mod c,ex]); 
+        if pos < l then
+          coc2 := [];
+          for j in [pos+1..l] do
+            a:=1;
+            h:=E(con)^A[5][i][y[j]-i];
+            for ex in [0..c] do
+              if a = h then
+                break;
+              fi;
+              a:=a*g;
+            od;
+            if ex=c then 
+             Print("\n The algebra is not a genuine cyclotomic algebra \n");
+              return fail;
+            fi;          
+            Add(coc2,ex); 
+          od;
+          Add(coc1,coc2);
+        fi;
+        pos := pos+1;
+      od;
+      if l=1 then 
+        return [A[1]*A[4][x][1],F,c,act[1]];
+      else 
+        return [A[1]*A[4][x][1],F,c,act,coc1];
+      fi;
+    fi;
+  fi;
+od;
+
+v := [1..k];
+d1 := List(v,i->Filtered(v,j->i=j or e[i,j]<>0));
+cls := [];
+
+w:=v;
+
+while w <> [] do
+  i:=w[1];
+  x:=[];
+  y:=[i];
+  while x<>y do
+    x:=y;
+    y:=Union(x,Concatenation(List(x,j->d1[j])));
+    w:=Difference(w,y);
+  od;
+  Add(cls,x);
+od;
+
+lc := Length(cls);
+
+if lc = 1 then 
+  return fail;
+fi;
+
+### WE CALCULATE ALL THE UNIONS OF CLASSES WITH AT LEAST TWO GENERATORS AND AT MOST HALF OF THE NUMBER OF GENERATORS
+
+gcls := Filtered(SSortedList(Arrangements(cls),Union),x->Size(x)>1 and 2*Size(x) <= k);
+SortBy(gcls,Size);
+
+for x in gcls do
+  y := Difference(v,x);
+  F1:=NF(con,List(y,j->A[4][j][2]));
+  F2:=NF(con,List(x,j->A[4][j][2]));
+  if IsCyclotomicExtension(F2,F) and IsCyclotomicExtension(F2,F) then    
```

Comment:
@drallenherman you can comment immediately in the diffs for the pull request - so, should the 1st `F2` be `F1` ?